src/manager/checkpoint.rs

```diff
@@ -0,0 +1,160 @@
+
+use crate::config::Subnet;
+use crate::lotus::client::LotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
+
+const CHAIN_HEAD_REQUEST_PERIOD: Duration = Duration::from_secs(10);
+
+/// Starts the checkpoint manager.
+#[allow(dead_code)]
+pub async fn start(subnets: HashMap<String, Subnet>) {
+    let managed_subnets = subnets_to_manage(subnets);
+
+    let mut futures = Vec::new();
+    for (child, parent) in managed_subnets {
+        futures.push(manage_subnet(child, parent));
+    }
+    join_all(futures).await;
+}
+
+/// This function takes a `HashMap<String, Subnet>` and returns a `Vec` of tuples of the form
+/// `(child_subnet, parent_subnet)`, where `child_subnet` is a subnet that we need to actively
+/// manage checkpoint for. This means that for each `child_subnet` there exists at least one account
+/// for which we need to submit checkpoints on behalf of to `parent_subnet`, which must also be
+/// present in the map.
+fn subnets_to_manage(subnets: HashMap<String, Subnet>) -> Vec<(Subnet, Subnet)> {
+    // First, we remap subnets by SubnetID.
+    let subnets_by_id: HashMap<SubnetID, Subnet> = subnets
+        .values()
+        .map(|s| (s.id.clone(), s.clone()))
+        .collect();
+
+    // Then, we filter for subnets that have at least one account and for which the parent subnet
+    // is also in the map, and map into a Vec of (child_subnet, parent_subnet) tuples.
+    subnets_by_id
+        .values()
+        .filter(|s| !s.accounts.is_empty())
+        .filter(|s| s.id.parent().is_some() && subnets_by_id.contains_key(&s.id.parent().unwrap()))
+        .map(|s| (s.clone(), subnets_by_id[&s.id.parent().unwrap()].clone()))
+        .collect()
+}
+
+/// Monitors a subnet `child` for checkpoint blocks. It emits an event for every new checkpoint block.
+async fn manage_subnet(child: Subnet, parent: Subnet) -> Result<()> {
+    let child_client = LotusJsonRPCClient::from_subnet(&child);
+    let parent_client = LotusJsonRPCClient::from_subnet(&parent);
+
+    // Read the parent's chain head and obtain the tip set CID.
+    let parent_head = parent_client.chain_head().await?;
+    let cid_map = parent_head.cids.last().unwrap().clone();
+    let parent_tip_set = Cid::try_from(cid_map)?;
+
+    // Extract the checkpoint period from the state of the subnet actor in the parent.
+    let state = parent_client
+        .ipc_read_subnet_actor_state(parent_tip_set)
+        .await?;
+    let period = state.check_period;
+
+    // We should have a way of knowing whether the validator has voted in the current open
+    // checkpoint epoch. For now, we maintain this information in memory and upon a restart only
+    // vote for the next epoch.
+
+    // We can now start looping. In each loop we read the child subnet's chain head and check if
+    // it's a checkpoint epoch. If it is, we construct and submit a checkpoint.
+    loop {
+        let child_head = child_client.chain_head().await?;
```

Comment:
I don't know if I like this check. What if you miss the epoch of the checkpoint when reading the head? I would suggest the following change. 
- Read the height of the head. 
- If head > epoch for the next checkpoint, see if we already voted for that checkpoint. 
- If this is not the case, submit the checkpoint. 
I think this may cover us against certain rough edges,

Reply:
Yes. Particularly, if we have the method you suggest above, it would make this relatively simple to implement.